### ISSUE TITLE
Stray drag-n-drop operation can easily destroy notebook structure behind user's back

### ISSUE THREAD

**Issue by pfalcon** [zim-desktop-wiki/zim-desktop-wiki#1443]
A few basic facts:

1. Zim's page drag-n-drop operations are not subject to undo.
2. The most popular computer form-factor now is a laptop, laptops use various "touchpads" as cursor control device, and quality of them vary wildly, many mis-register normal movement as drag, etc.

This makes drag-n-drop in Zim truly a destructive operation. Just yesterday I faced a few cases (many a single day, definitely faced it before): a) the whole page sub-hierarchy got under another page, and apparently stayed that way for weeks unnoticed; b) in tag filter view (which also has drag-n-drop enabled!), a day's page got moved under completely different day's page, leading to a monster like "Journal:2021:03:28:21". The original page also happened to not have extra date header in it, which made it almost impossible to reposition back.

**Comment by pfalcon** [comment]
There're 2 possible solutions to this issue:

1. Disable drag-n-drop in page tree/list altogether. There's still "Rename or move page..." command, which is easy enough to use thanks to pagename completion in the dialog, but otherwise enough handwork to make users think what they're actually doing.
2. Don't perform drag-n-drop operation right away, but show a confirmation dialog, asking users to think about effects of the operation which was requested. This approach is used e.g. in the Claws Mail mail client, where it's similarly easy to move a mail subfolder around, which is otherwise expensive and not easily undoable operation.

In my fork, https://github.com/pfalcon/zim-desktop-wiki-markdown / https://github.com/zim-desktop-wiki/zim-desktop-wiki/issues/26#issuecomment-770992975 , I'm going to try to go with p.2, short of completely disabling it (p.1).

**Comment by pfalcon** [comment]
Well, I now checking that even with "Notebook Editable" unchecked (i.e. notebook read-only), it's possible to drag-n-drop pages.

**Comment by pfalcon** [comment]
Just to back this report with related cases where mis-pressing something in Zim (which happens a lot to us humans!) leads to loads of user confusion and frustration: https://github.com/zim-desktop-wiki/zim-desktop-wiki/issues/1007 . That's poor UX. If a notebook is read-only, trying to edit it (pages, or structure) should lead to dialog popping up telling them something like "You cannot edit this notebook because it's read-only."

**Comment by pfalcon** [comment]
Experimenting more with this, I found that there're more issues. E.g., if there's a parent and child page, it's possible to drag-n-drop parent onto child. In this kind, parent-child hierarchy is kinda duplicated (i.e. it becomes parent-child-parent-child). But trying to remove leaf child actually seems to lead to the removal of the original child (and whole subtree). So again, drag-n-drop in Zim is confusing, borders on data loss, and should be gated behind confirmation (I'm optimistic here of course, I love features. A reasonable person would say "should be disabled").

**Comment by mnordlin** [comment]
The "Stray drag-n-drop operation" is indeed a major problem to me also. It happens a lot, and takes forever to rectify. I'd say it happens a lot more often with the later-day UX as the movable ribbon between the panels has shrunk and is now minimal. I looked around for a setting to disable drag-n-drop in page tree/list altogether, but found none. Please, please find a way to disable drag-n-drop at least for the Windows version.

**Comment by dreua** [comment]
I on the other hand was very confused, that drag and drop was not working any more after an a minor version upgrade from 0.74.1 to 0.74.2. (Took me a `git diff 0.74.1...0.74.2 -- . ':!translations'` to figure this out, not something for the end user ;))

**Comment by oturpe** [comment]
> I on the other hand was very confused, that drag and drop was not working any more after a patch version upgrade from 0.74.1 to 0.74.2. (Took me a `git diff 0.74.1...0.74.2 -- . ':!translations'` to figure this out, not something for the end user ;))
> 
> I've never experienced issues with drag and drop and use it quite often. Maybe we can fix this in a way that is safe and usable?

Drag&drop was not completely removed.
It was made optional.
You can enable it at
_Edit_ → _Preferences_ → _Plugins_ → _Page Index_ → _Configure_ ,
option _Use drag&drop to move pages in the notebook_.

**Comment by dreua** [comment]
> Drag&drop was not completely removed.
> It was made optional.

Sorry, this is what I meant. To be even more exact, it was made optional *and* disabled by default. Which looked a lot like a bug to me, not like an intended change.

**Comment by martinstut** [comment]
> it was made optional _and_ disabled by default. Which looked a lot like a bug to me, not like an intended change.

So this looks like a documentation issue. Where would people look for help if they suddenly (e.g. after upgrading or installing a fresh version on a new computer) get desperate about drag & drop no longer working? IMO this should be at least an FAQ entry.

As a long time Zim user, I stumbled across this and it significantly reduced my use of Zim. Drag & Drop rearrangement was what I liked most with Zim (even using it as a Kanban-like project tracking tool).

**Comment by sojusnik** [comment]
Yeah, it's very confusing indeed, even for a long time Zim user like me, f.i. drag-n-drop still wasn't reimplemented for the tag side pane, only for the index pane, see [here](https://github.com/zim-desktop-wiki/zim-desktop-wiki/issues/1781), so I also thought for a long time that this is a bug.

**Comment by introt** [comment]
Imo a good middle-ground between discoverability and usability would be to have the index switch into a "refactoring mode"[^link] on drag-n-drop, highlighting the affected pages & requiring the user to click a confirmation button (or pressing a keyboard shortcut) before actually executing the destructive page moves. This approach would also allow prompting the user about link updating preferences only once when moving multiple pages.

[^link]: previously discussed over at https://github.com/zim-desktop-wiki/zim-desktop-wiki/discussions/2015#discussioncomment-2619687

**Comment by yareckon** [comment]
Oh man, I just found this.  This decision to turn this off by default wasted weeks of my time migrating my content away from Zim over the last year.  Why?  Well I came to zim last year, but left it for another app because of the lack of page reordering.  I copied everything over manually, and have put up with a poorer experience since, simply to be able to create an index without naming pages 01, 02 etc...  Just so you know, because this is off **as far as people trying the software can see, Zim can't reorder pages at all**.  They are not going to poke around in an impossible to find plugin config and notice an unchecked checkbox, they are just going to try another note app that **has** this basic feature. Defaults should continue previous behaviors or be unmissably documented!
